vql/tools/sftp_upload.go

```diff
@@ -0,0 +1,202 @@
+//+build extras
+
+package tools
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net"
+	"path/filepath"
+	"strings"
+
+	"github.com/Velocidex/ordereddict"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/net/context"
+	"www.velocidex.com/golang/velociraptor/file_store/api"
+	"www.velocidex.com/golang/velociraptor/glob"
+	vql_subsystem "www.velocidex.com/golang/velociraptor/vql"
+	"www.velocidex.com/golang/vfilter"
+)
+
+type SFTPUploadArgs struct {
+	File       string `vfilter:"required,field=file,doc=The file to upload"`
+	Name       string `vfilter:"optional,field=name,doc=The name of the file that should be stored on the server"`
+	User       string `vfilter:"required,field=user,doc=The username to connect to the endpoint with"`
+	Path       string `vfilter:"required,field=path,doc=Path on server to upload file to"`
+	Accessor   string `vfilter:"optional,field=accessor,doc=The accessor to use"`
+	PrivateKey string `vfilter:"required,field=privatekey,doc=The private key to use"`
+	Endpoint   string `vfilter:"required,field=endpoint,doc=The Endpoint to use"`
+	HostKey    string `vfilter:"optional,field=hostkey,doc=Host key to verify. Blank to disable"`
+}
+
+type SFTPUploadFunction struct{}
+
+func (self *SFTPUploadFunction) Call(ctx context.Context,
+	scope *vfilter.Scope,
+	args *ordereddict.Dict) vfilter.Any {
+
+	arg := &SFTPUploadArgs{}
+	err := vfilter.ExtractArgs(scope, args, arg)
+	if err != nil {
+		scope.Log("upload_sftp: %s", err.Error())
+		return vfilter.Null{}
+	}
+
+	err = vql_subsystem.CheckFilesystemAccess(scope, arg.Accessor)
+	if err != nil {
+		scope.Log("upload_SFTP: %s", err)
+		return vfilter.Null{}
+	}
+
+	accessor, err := glob.GetAccessor(arg.Accessor, scope)
+	if err != nil {
+		scope.Log("upload_SFTP: %v", err)
+		return vfilter.Null{}
+	}
+
+	file, err := accessor.Open(arg.File)
+	if err != nil {
+		scope.Log("upload_SFTP: Unable to open %s: %s",
+			arg.File, err.Error())
+		return &vfilter.Null{}
+	}
+	defer file.Close()
+
+	if arg.Name == "" {
+		arg.Name = arg.File
+	}
+
+	stat, err := file.Stat()
+	if err != nil {
+		scope.Log("upload_SFTP: Unable to stat %s: %v",
+			arg.File, err)
+	} else if !stat.IsDir() {
```

Comment:
Maybe structure this code like:

```
if !stat.IsDir {
return &vfilter.Null{}
}
```

Since we can not really upload a directory no point continuing on here.